cmd/kubeadm/app/cmd/join.go

```diff
@@ -245,27 +296,181 @@ func NewJoin(cfgPath string, args []string, defaultcfg *kubeadmapiv1alpha3.JoinC
+		glog.V(1).Infoln("[join] retrieving KubeConfig objects")
+		masterConfiguration, err := j.FetchInitClusterConfiguration(tlsBootstrapCfg)
+		if err != nil {
+			return err
+		}
+
+		// injects into the kubeadm configuration used for init the information about the joining node
+		masterConfiguration.NodeRegistration = j.cfg.NodeRegistration
+		masterConfiguration.API.AdvertiseAddress = j.cfg.AdvertiseAddress
+
+		// Checks if the cluster configuration supports
+		// joining the additional master node and if all the necessary certificates are provided
+		if err = j.CheckIfReadyForJoinAsMaster(masterConfiguration); err != nil {
+			// outputs the not ready for join master message
+			ctx := map[string]string{
+				"Error": err.Error(),
+			}
+
+			var msg bytes.Buffer
+			notReadyToJoinMasterTemp.Execute(&msg, ctx)
+			return errors.New(msg.String())
+		}
+
+		// run init master preflight checks for checking all the prequisites
+		glog.Infoln("[join] running pre-flight checks before initializing the new master node")
+		preflight.RunInitMasterChecks(utilsexec.New(), masterConfiguration, j.ignorePreflightErrors)
+
+		// Prepares the node for joining as a master by writing necessary
+		// KubeConfig files, and static pod manifests
+		if err = j.PrepareForJoinAsMaster(masterConfiguration); err != nil {
+			return err
+		}
+	}
+
+	// Executes the kubelet TLS bootstrap process, that completes with the node
+	// joining the cluster with a dedicates set of credentials as required by
+	// the node authorizer.
+	// if the node is joining as a master, since it uses static pods for the control plane,
+	// as soon as the kubelet starts it will take charge of creating control plane
+	// components on the node.
+	if err = j.BootstrapKubelet(tlsBootstrapCfg); err != nil {
+		return err
+	}
+
+	// if the node is joining as a master
+	if j.cfg.Master == true {
+		// Marks the node with master taint and label.
+		if err := j.MarkMaster(); err != nil {
+			return err
+		}
+
+		// outputs the join master done template and exits
+		ctx := map[string]string{
+			"KubeConfigPath": kubeadmconstants.GetAdminKubeConfigPath(),
+		}
+		joinMasterNodeDoneTemp.Execute(out, ctx)
+		return nil
+	}
+
+	// otherwise, if the node joined as a worker node;
+	// outputs the join done message and exits
+	fmt.Fprintf(out, joinWorkerNodeDoneMsg)
+	return nil
+}
+
+// FetchInitClusterConfiguration reads the master configuration from the kubeadm-admin configMap,
+func (j *Join) FetchInitClusterConfiguration(tlsBootstrapCfg *clientcmdapi.Config) (*kubeadmapi.InitConfiguration, error) {
+	// creates a client to access the cluster using the bootstrap token identity
+	tlsClient, err := kubeconfigutil.ToClientSet(tlsBootstrapCfg)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to access the cluster: %v", err)
+	}
+
+	// Fetches the master configuration from the cluster
+	kubeadmConfig, err := configutil.FetchConfigFromFileOrCluster(tlsClient, os.Stdout, "join", "")
```

Comment:
We should make it clear that we don't support version skew joins for new feature additions.

Reply:
I'm not sure I got your comment right, but since we are reading the cluster configuration, there is no chance to have version skew